### PR TITLE
Adding loader for OTMM makam recognition dataset (CompMusic corpus)

### PR DESCRIPTION
**Pull request by genisplaja**
### Adding loader for OTMM makam recognition dataset (CompMusic corpus)

This is one of the most complete and large datasets for makam recognition. This dataset was compiled within the framework of the CompMusic project (like Saraga Carnatic and Hindustani).

### Description

Please include the following information at the top level docstring for the dataset's module mydataset.py:

- [x] Describe annotations included in the dataset
- [x] Indicate the size of the datasets (e.g. number files and duration, hours)
- [x] Mention the origin of the dataset (e.g. creator, institution)
- [x] Describe the type of music included in the dataset
- [x] Indicate any relevant papers related to the dataset

#### Dataset loaders checklist:

- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`, which generates an index file.
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`.
- [x] Create a module in mirdata, e.g. `mirdata/my_dataset.py`
- [x] Create tests for your loader in `tests/`, e.g. `test_my_dataset.py`
- [x] Add your module to `docs/source/mirdata.rst` and `docs/source/quick_reference.rst`
- [x] Run `tests/test_full_dataset.py` on your dataset.

### CODE REVIEW

File: tests/test_otmm_makam.py
Comment (rabitt) [inline tests/test_otmm_makam.py:112]: just a nit, and up to you - for the the tests, you dont have to use the complete real data files - you can reduce them so they have e.g. just one item in each dictionary, and it will make the test code smaller

**Review by rabitt — CHANGES_REQUESTED** [review]
This looks great @genisplaja ! There are a couple added files that got checked in that shouldn't be there (acousticbrainz_genre_dataset_little_test and a .gitignore file), and I made a couple of small comments.

Depending on which gets merged first, note that there are a few more small (simplifying) changes in #443 that will affect this PR. 

New loader checklist:
- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`
- [x] Create a module in `mirdata`, e.g. `mirdata/my_dataset.py`
- [x] Create tests in `tests/`, e.g. `test_my_dataset.py`
- [x] Add the module to `docs/source/mirdata.rst`
- [x] Run test_full_dataset.py
- [ ] Make sure the docs build properly

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:185]: Sure, I'll do that!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (nkundiushuti) [inline mirdata/datasets/compmusic_otmm_makam.py:2]: should this have a method to allow you to work with private audio?

**Review by nkundiushuti — APPROVED** [review]
why did you include that index for acousticbrainz

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:254]: Thanks:)

**Comment by genisplaja** [comment]
> why did you include that index for acousticbrainz

Sorry, my bad! Removed it in this last commit.

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (rabitt) [inline mirdata/datasets/compmusic_otmm_makam.py:196]: nit - this function can simply `return json.load(fhandle)`!

Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:196]: True! Fixed now:)

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (rabitt) [inline mirdata/datasets/compmusic_otmm_makam.py:174]: Just to double check, can you confirm that 1) the hop size is exactly 0.0029 s, and that the frames are left aligned, and not center aligned? In the mirdata paper, this was one of the big sources of differences when evaluating on ikala - slight differences in the hop size and in the alignment of the time stamps

cc @nkundiushuti

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (rabitt) [inline mirdata/datasets/compmusic_otmm_makam.py:229]: It would be best to use a different variable name than `track` here, since it's also an attribute of `Dataset` and it could cause unexpected bugs.

Comment (rabitt) [reply mirdata/datasets/compmusic_otmm_makam.py:229]: that said, I guess it won't actually cause bugs since `self.track` != `track`, but just for clarity!

File: tests/test_otmm_makam.py
Comment (genisplaja) [reply tests/test_otmm_makam.py:112]: yes, I am aware I did the same for the Saraga dataset! In this case, I noticed that there were a lot of special characters in the metadata files so I thought it wouldn't do bad to check the encoding/decoding of the json file. But I'll keep that in mind!!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py]: Ok! Just changed for the regular `i` so we avoid any kind of confusion.

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:173]: Yes! The dataset's README states the following : _the timestamps are removed to reduce the filesizes. The step size of the pitch track is 0.0029 seconds (an analysis window of 128 sample hop size of an mp3 with 44100 Hz sample rate), with which one can recompute the timestamps of samples._ And so I did!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:2]: Hmm.. that's a good question! In the Zenodo repo (https://zenodo.org/record/58413#.YBLCaWPOH-b) it is said that the audio cannot be shared (maybe it cannot even be shared for research purposes), so I am not sure if there is a source from where to get the audio tracks! That would be quite essential for the index to be properly written. Should we ask Xavier and the rest of the compilers?

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (rabitt) [reply mirdata/datasets/compmusic_otmm_makam.py:173]: Ah great! if that's the case, can we use `time_step = 128 / 44100`? It's very close to 0.0029, but over time the small difference propagates! And we can remove the `round`.

On the frame alignment, it looks like it's not clear from the documentation. Maybe we should check it with the dataset creators? Center aligned frames are more likely to be the "correct" way to align it, but maybe not!

**Comment by sertansenturk** [comment]
This is a nice surprise! Thanks, @nkundiushuti for letting me know!

If the PR does not have an extremely tight schedule, I would be happy to give my review (most likely this weekend). 

Apart from https://mirdata.readthedocs.io/en/latest/source/contributing.html, are there any guidelines/styles, etc. would you like me to adhere to?

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (sertansenturk) [reply mirdata/datasets/compmusic_otmm_makam.py:173]: 💯 for `time_step = 128 / 44100 seconds`, 0.0029 is not clear.

f0 was computed using [this tool](https://github.com/sertansenturk/predominantmelodymakam/tree/f8b7302bc657f90e2b10a0ffd988902935adc3d6), which basically wraps Essentia. In essence, we should check how [FrameGenerator](https://essentia.upf.edu/reference/std_FrameGenerator.html) dealt with alignment.

The current doc specifies centering as:

```
startFromZero (bool ∈ {true, false}, default = false) :
   whether to start the first frame at time 0 (centered at frameSize/2) if true, or -frameSize/2 otherwise (zero-centered)
```

However, we may want to ensure that the algorithm worked the same circa 2016. 
We seem to have noted the Essentia version as ([commit: e2339c](https://github.com/MTG/essentia/commit/e2339c11f869a850fc15a3fff7a75264db92fe11)) in this [JSON file](https://github.com/MTG/otmm_makam_recognition_dataset/blob/dlfm2016/algorithms.json).

@genisplaja if you can confirm the answer from Essentia's source code by yourself, I'd be glad. Otherwise, I'll dig more over the weekend.

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (sertansenturk) [reply mirdata/datasets/compmusic_otmm_makam.py:2]: The audio is accessible through [pycompmusic](https://github.com/MTG/pycompmusic) given that to researchers, who are given access to Dunya upon [registration](https://dunya.compmusic.upf.edu/user/register/).

I don't know if `mirdata` provides (or should provide) an adapter to connect to such external sources via security tokens. Either way, I think it's between the current MTGers and `mirdata` maintainers to give a decision. If there is appetite, I can share some starter code to download the relevant audio.

File: tests/test_otmm_makam.py
Comment (sertansenturk) [reply tests/test_otmm_makam.py:112]: Then it might be better to have concise unittest(s) for characters specific to Turkish, namely çÇ, ğĞ, ıİ, iİ, öÖ, şŞ, üÜ. (Emphasis on the difference in the capitals of ı and I.)

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (alastair) [reply mirdata/datasets/compmusic_otmm_makam.py:2]: hey @sertansenturk, we're already working with @genisplaja to integrate pycompmusic into mirdata, so this would definitely be a possible way to load the data.

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:173]: Thanks @rabitt for the suggestion, I'll fix it right now.

Sure @sertansenturk, thank you very much for all the information! I've gone through both implementations and it looks like the algorithm hasn't changed at all, and the default parameters are configured the same, so the frames should be zero-centered!

File: tests/test_otmm_makam.py
Comment (genisplaja) [reply tests/test_otmm_makam.py:112]: I'll introduce all these characters to the test file and see if there's any problem showing up. Thanks for the suggestion!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (sertansenturk) [reply mirdata/datasets/compmusic_otmm_makam.py:2]: Neat! Then, I'll stop trying to sell sawdust to the lumber mill! :stuck_out_tongue_closed_eyes:

File: tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/metadata.py
Comment (sertansenturk) [inline tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/metadata.py:107]: Please remove the stray debugger breakpoint.

File: tests/test_otmm_makam.py
Comment (sertansenturk) [inline tests/test_otmm_makam.py:9]: Echoing on @rabitt's [comment](https://github.com/mir-dataset-loaders/mirdata/pull/435#r566016550_) regarding the length of unit tests. It's a bit difficult to capture what each test at a glimpse due to their size. 

The ideal approach might be to use [mocks](https://docs.python.org/3/library/unittest.mock.html) to decouple the functionality you are testing from the data. On the other hand, I appreciate that mocks are a bit difficult to get your head around. Please disregard my suggestion if you think it's overkill.

File: tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/wrapper_melody.py
Comment (sertansenturk) [inline tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/wrapper_melody.py:8]: Is this useful to port this script to `mirdata`?

File: tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/metadata.py
Comment (sertansenturk) [inline tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/metadata.py:1]: Is this useful to port this script to `mirdata`?

**Review by sertansenturk — COMMENTED** [review]
Great work @genisplaja et al! It's so nice to see the dataset is being integrated. I guess I'll load the dataset using `mirdata`, if I ever get to work on makam recognition again!

Had a chance to do a quick review. It's mostly nitpicky so please feel free to ignore them if they are not meaningful. I'll do a proper review over the weekend.

P.S: It must have been such a pain to go through the rookie Python mistakes we did 5 years ago, kudos to you!

File: tests/test_otmm_makam.py
Comment (sertansenturk) [inline tests/test_otmm_makam.py:480]: This should be a separate test.

**Review by sertansenturk — CHANGES_REQUESTED** [review]
(Finalizing my review)

Great work! :100: :clap: 

I installed the branch and played with it. I can confirm that everything works as intended. You can find a summary at the end of the review.

I only have two minor change requests in `compmusic_otmm_makam.py` (one linting, one code simplification). The rest can be taken as suggestions, and you can ignore them if you'd like.

A few additional points below:

- I made a [minor fix](https://github.com/sertansenturk/otmm_makam_recognition_dataset/tree/dlfm2016-fix1) to the dataset a few months ago, which has not been pushed to Zenodo yet. I think it's a separate conversation between MTG'ers (@nkundiushuti, @alastair) and me, but I wanted to note it here.
- In the codebase, the dataset and the relevant methods/variables, etc., are either called "OTMM makam" or "Compmusic OTMM makam." It's nothing dire, but I wanted to point out if the maintainers are extra-careful with consistency.
- It might be convenient to use `pathlib` instead of `os.path` in the future.
- I noticed that tests have multiple asserts. I think this pattern is OK, and since the tests pass, there is no need to change anything. However, it might be desirable to convert the assertions to [subtests](https://docs.python.org/3/library/unittest.html#subtests) in the future, so that a test will not stop after the first failure.

## Summary of my checks

0. Get [Docker](https://docs.docker.com/get-docker/)
1. Create a Dockerfile

```
FROM python:3.8

RUN apt-get -qq update && \
    apt -qq install -y libsndfile1 && \
    rm -rf /var/lib/apt/lists/* && \
    pip install --upgrade pip && \
    git clone https://github.com/mir-dataset-loaders/mirdata.git /usr/src/mirdata/

WORKDIR /usr/src/mirdata

RUN git checkout genis/OTMM_makam_recognition && \
    pip install . && \
    pip install .[tests] && \
    pip install .[docs] && \
    pip install .[dali]
    
# run only makam tests
# Note: Some out-of-scope tests fail when we run all tests
RUN pytest ./tests/test_otmm_makam.py --local

```

In terminal run;

```bash
docker build . -t mirdata/otmm && docker run --rm -it -v "$PWD":/usr/src/mirdata mirdata/otmm
```

Run the Python commands:

```python
import mirdata
makamset = mirdata.initialize('compmusic_otmm_makam')
makamset.download()
makamset.validate()
example_track = makamset.choice_track()
print(example_track)
example_track.pitch.frequencies
example_track.pitch.frequencies.size
```

Everything worked as intended :partying_face:

**Comment by magdalenafuentes** [comment]
Wow @sertansenturk thanks for the detailed review! 🥇

**Comment by genisplaja** [comment]
Thank you very much for the review @sertansenturk. I really appreciate it! I will follow it carefully ASAP, and discuss some of the proposed consistency issues in the next meeting with all the _mirdata_ team. Thanks again and feel free to re-review if needed after the changes are made.

File: tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/wrapper_melody.py
Comment (genisplaja) [reply tests/resources/mir_datasets/compmusic_otmm_makam/MTG-otmm_makam_recognition_dataset-f14c0d0/extras/wrapper_melody.py:8]: It won't be used for the mirdata loader, but since the whole set of dataset files is compressed in a single .zip file, when retrieving the data from Zenodo this script file is also downloaded and extracted from the .zip file (same happens for `metadata.py`. I decided not to write an extra download function to delete this script once downloaded (it may even be useful for the dataset users). However, it was my bad to add it into the testing resources, because it is not needed and it is unnecessarily occupying space. Same for `metadata.py`. Will remove them, thanks for the catch!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:179]: Sure! It was coded like this but I changed it because of the bug mentioned in #449. Since this is fixed, I will recode it in just these three lines.

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:48]: Sure!! My name is also written with í (Genís), and it is quite important for the name pronunciation. However, I literally copied this citation from Zenodo! Maybe we could fix it in the root citation as well. Thanks!

Comment (sertansenturk) [reply mirdata/datasets/compmusic_otmm_makam.py:48]: Thanks for the warning! I fixed the citation in Zenodo!

**Comment by genisplaja** [comment]
By the way @sertansenturk, about the first point you mentioned: _I made a minor fix to the dataset a few months ago, which has not been pushed to Zenodo yet. I think it's a separate conversation between MTG'ers (@nkundiushuti, @alastair) and me, but I wanted to note it here._ You think this fix would affect the loader and its implementation if pushed to Zenodo as well? I am asking because if at some point this fix is updated in the Zenodo repository, it would be good to know if this would affect the loader. Thank you very much!

**Comment by sertansenturk** [comment]
> By the way @sertansenturk, about the first point you mentioned: _I made a minor fix to the dataset a few months ago, which has not been pushed to Zenodo yet. I think it's a separate conversation between MTG'ers (@nkundiushuti, @alastair) and me, but I wanted to note it here._ You think this fix would affect the loader and its implementation if pushed to Zenodo as well? I am asking because if at some point this fix is updated in the Zenodo repository, it would be good to know if this would affect the loader. Thank you very much!

Good call! Referring to the [diff between MTG/otmm_makam_recognition_dataset and sertansenturk/otmm_makam_recognition_dataset](https://github.com/MTG/otmm_makam_recognition_dataset/compare/master...sertansenturk:master), there is nothing major. 

Having said that there can be a divergence between MusicBrainz, Dunya & the dataset when a source merges a UUID (see [the only example between lines 180-184 annotations.json](https://github.com/MTG/otmm_makam_recognition_dataset/compare/master...sertansenturk:master#diff-2e28a4f50d7745d2df9f73ba128a92355cacd2ca9f4883fad4495ff93d657ea6R392)). Basically when/if the loader requests audio recordings, it should fetch from the `dunya_uid` if key exists else from `mbid`.  Since the loader does not deal with audio at the moment, nothing changes.

**Comment by genisplaja** [comment]
Hey @rabitt! I think there is still an issue with mypy checks since I can see that they are failing also for `cante100`, `salami`, `medley_db` and `beatles` loaders. I have updated the branch to the latest master changes but it's still breaking for that reason...!

**Comment by genisplaja** [comment]
> Having said that there can be a divergence between MusicBrainz, Dunya & the dataset when a source merges a UUID (see [the only example between lines 180-184 annotations.json](https://github.com/MTG/otmm_makam_recognition_dataset/compare/master...sertansenturk:master#diff-2e28a4f50d7745d2df9f73ba128a92355cacd2ca9f4883fad4495ff93d657ea6R392)). Basically when/if the loader requests audio recordings, it should fetch from the `dunya_uid` if key exists else from `mbid`. Since the loader does not deal with audio at the moment, nothing changes.

Cool! Thank you for the answer @sertansenturk. As soon as the audio can be accessed through mirdata and the Dunya API, we will review the OTMM Makam Dataset loader to implement the audio support.

**Comment by rabitt** [comment]
> Hey @rabitt! I think there is still an issue with mypy checks since I can see that they are failing also for `cante100`, `salami`, `medley_db` and `beatles` loaders. I have updated the branch to the latest master changes but it's still breaking for that reason...!

Hey @genisplaja yes sorry it wasn't clear from the opened issue - I fixed it in #443 which I just merged. If you update your branch it should fix the problem!

**Comment by rabitt** [comment]
@genisplaja FYI I'm going to push a small commit to this branch to get it up to date with #443 .

**Comment by rabitt** [comment]
Related to Sertan's points about consistency - do you prefer to call this dataset "compmusic_otmm_makam" or "otmm_makam"? Since #443 , we're more strict with the name (the index name has to match) and right now it's not consistent. If it's up to me I'd call it the shorter one (otmm_makam) but no problem if we want to add compmusic in front. Just let me know!

**Comment by genisplaja** [comment]
> @genisplaja FYI I'm going to push a small commit to this branch to get it up to date with #443 .

Ups sorry, haven't seen this message. I hope that I haven't messed up what you were working on. Thanks!

**Comment by rabitt** [comment]
> > @genisplaja FYI I'm going to push a small commit to this branch to get it up to date with #443 .
> 
> Ups sorry @rabitt, haven't seen this message. I hope that I haven't messed up what you were working on. Thanks!

No problem! Can you let me know your choice about the dataset name?

**Comment by genisplaja** [comment]
> No problem! Can you let me know your choice about the dataset name?

Sure @rabitt! We have discussed that at the end of the meeting this afternoon and we have decided to keep the `compmusic` in the dataset name. I was proposing not to add `compmusic` in the index, script and test files, since you can still relate them and long filenames are avoided. But I understand the consistency point here, so I would keep `compmusic` everywhere rather than remove it.

**Comment by rabitt** [comment]
@genisplaja I think this is good to go! Let me know if it's OK to merge (or feel free to click it yourself :) )

**Comment by genisplaja** [comment]
Thanks @rabitt and everyone for the reviews, comments, and suggestions! I'll take a final look at it myself today, update the branch and let you know when it's ready!

**Comment by genisplaja** [comment]
Ok! I reviewed and adapted the loader to the last updates in master. For me, it is good to go:) I'll wait until the afternoon in case somebody has any additional comment or request, otherwise I will merge it myself, or feel free to merge @rabitt. Thanks!

**Review by sertansenturk — APPROVED** [review]
lgtm

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (rabitt) [inline mirdata/datasets/compmusic_otmm_makam.py:78]: tiny nit - pitch_paths and mb_tags_paths are missing from this docstring

File: mirdata/jams_utils.py
Comment (rabitt) [inline mirdata/jams_utils.py:37]: good catch!

**Comment by rabitt** [comment]
@genisplaja I made one last comment, but feel free to do the merging honors!

File: mirdata/datasets/compmusic_otmm_makam.py
Comment (genisplaja) [reply mirdata/datasets/compmusic_otmm_makam.py:78]: True!!!! Sorry, haven't noticed. I'll fix it now.
By the way, `tonic`, `makam` and `mbid` are properties now. Should I add a `Properties` section in the docstring and put them there?
```
Properties:
    tonic (float): bla bla bla
    makam (str): bla bla bla
    ...
Catched properties:
    pitch (F0Data): bla bla bla
```

**Pull request merged**
